torchinductor/scheduler.py

```diff
@@ -1000,188 +761,279 @@ def dead_node_elimination(self):
+                    for n in node.recursive_predecessors - combined_predecessors
                 )
+            return False
+
+        visited = set()
+        combined_names = node1.get_names() | node2.get_names()
+        combined_predecessors = (
+            node1.recursive_predecessors | node2.recursive_predecessors
+        ) - combined_names
+        return any(check(self.name_to_fused_node[n]) for n in combined_predecessors)
+
+    def can_fuse(self, node1: BaseSchedulerNode, node2: BaseSchedulerNode):
+        """
+        Determine if it is possible to combine node1 and node2 into a
+        single fused node.
+        """
+        if node1 is node2:
+            return False
+        if isinstance(node1, (ExternKernelSchedulerNode, NopKernelSchedulerNode)):
+            return False
+        if isinstance(node2, (ExternKernelSchedulerNode, NopKernelSchedulerNode)):
+            return False
+        if node2.get_names() & node1.recursive_predecessors:
+            return False  # node2 must go before node1
+        if node2.is_template():
+            return False  # only epilogues
+
+        device = node1.get_device()
+        if device != node2.get_device():
+            return False  # wrong device
+
+        no_shared_data = self.score_fusion_memory(node1, node2) == 0
+        if no_shared_data and (
+            not config.aggressive_fusion or node1.is_reduction() or node2.is_reduction()
+        ):
+            return False  # heuristic not needed for correctness
+
+        if node1.get_names() & node2.recursive_predecessors:
+            # node2 depends on node1 outputs
+            if not self.can_fuse_vertical(node1, node2):
+                return False
+            if node1.is_template():
+                return template_can_fuse(node1, node2)
+            return self.get_backend(device).can_fuse_vertical(node1, node2)
+        else:  # nodes don't depend on each other, but may have common reads
+            if node1.is_template():
+                return False
+            return self.get_backend(device).can_fuse_horizontal(node1, node2)
+
+    def can_fuse_vertical(self, node1, node2):
+        """
+        Check if it is legal to fuse a consumer (node2) into a producer (node1).
+
+        We can fuse them if all the reads of node2 either match
+        corresponding writes in node1, or are written by nodes that can
+        be scheduled before the fusion of node1 and node2.
+        """
+        node1_names = node1.get_names()
+        remaining_deps = {
+            dep.name for dep in node2.unmet_dependencies - node1.read_writes.writes
+        }
+        if remaining_deps & node1_names:
+            return False  # MemoryDeps didn't match
+        for name in remaining_deps:
+            if node1_names & self.name_to_fused_node[name].recursive_predecessors:
+                return False
+        return True
+
+    def score_fusion(self, node1: BaseSchedulerNode, node2: BaseSchedulerNode):
```

Comment:
This is the key function that decides what order fusions are done in.  cc @ngimel on if this heuristic makes sense.